### PR TITLE
Fix #900, add inline whitelisting capabilities

### LINKED ISSUE #900: Add per-line whitelisting capabilities
We currently support a few different means of whitelisting: `--exclude`, `--exclude-dir`, and `pattern-not` to some extent. These are useful, but don't have the level of precision an AppSec team may need. Often a security engineer will want to whitelist a specific line in a file, not the entire file. Let's add functionality around whitelisting a specific line in a file. For example:

```python
os.system('ls -alh')  # noqa command-injection-os-system: constant argument is okay
```

A good first pass would be simply including something like `noqa` as a comment to whitelist a line. A more fully featured version would be something like:

```
noqa <rule-id>: <message>
```

Where `<rule-id>` is optional and will only whitelist the specific rule. A naked `noqa` will whitelist all rules. The `<message>` is also optional and is used to describe _why_ it's whitelisted. A common line will be something like `noqa command-injection-os-system: AppSec reviewed`.

Here are how some other linter/static analysis apps do it:

* `flake8`: https://flake8.pycqa.org/en/latest/user/violations.html - they have a very comprehensive story for selecting and ignoring/whitelisting.
* `bandit`: https://bandit.readthedocs.io/en/latest/config.html#suppressing-individual-lines
* `eslint`: https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments - they also have a good story here.

### PR DESCRIPTION
**Pull request by mschwager**

### CODE REVIEW

File: semgrep/semgrep/rule_match.py [hunk 1]
```diff
@@ -76,7 +76,7 @@ def end(self) -> Dict[str, Any]:
         return self._end
 
     @property
-    def lines(self) -> Iterable[str]:
+    def lines(self) -> List[str]:
```
Comment (mschwager) [inline semgrep/semgrep/rule_match.py:79]: This was necessary to perform list indexing in `rule_match_nosem`. The `typing` docs also suggest using `List` for return types and `Iterable` for arguments: https://docs.python.org/3/library/typing.html#typing.List

**Review by mschwager — COMMENTED** [review]
I decided to punt on `nosem` inline messages for now. E.g. `# nosem: rule-id <explanatory-comment>`. We can iterate on this if we find that functionality necessary.

**Review by brendongo — APPROVED** [review]

**Comment by dlukeomalley** [comment]
@mschwager @brendongo Can we add nosem tests for all of our supported languages? Python, Java, JavaScript, Golang, C?

**Comment by dlukeomalley** [comment]
@mschwager What were the pros/cons of choosing `nosem` vs. `noqa` or `nosec`?

**Comment by mschwager** [comment]
> @mschwager @brendongo Can we add nosem tests for all of our supported languages? Python, Java, JavaScript, Golang, C?

Sure thing, I created #1045 to track this.

> @mschwager What were the pros/cons of choosing `nosem` vs. `noqa` or `nosec`?

Good question! Initially I was planning on using `noqa`. While this would work seamlessly with people used to `flake8` it would also likely unintentionally ignore `semgrep` findings. For example, if there's an existing `flake8` finding on a line, a developer adds a naked `noqa` to ignore it, then `semgrep` finds something on the same line it will also be ignored.

As for the string itself, I didn't put too much thought into it. `eslint` throws brevity to the wind and uses `eslint-ignore`, which I'm not necessarily opposed to. Something like `semgrep-ignore` is very descriptive, but I went with something short and sweet.

**Comment by dlukeomalley** [comment]
Thanks for the ticket and thoughts, they're both helpful. A fun riff for us to consider: `nosg` ;)

**Comment by underyx** [comment]
"nem" is how we say "no" in Hungarian, so if it were up to me, I'd just go with `# nemgrep` 😛

**Comment by mschwager** [comment]
> "nem" is how we say "no" in Hungarian, so if it were up to me, I'd just go with `# nemgrep` stuck_out_tongue

`nemsem` it is!

**Comment by underyx** [comment]
`nemsem` means "not even" :D

**Pull request merged**